src/librustc_ast_passes/ast_validation.rs

```diff
@@ -96,6 +99,27 @@ impl<'a> AstValidator<'a> {
         self.bound_context = old;
     }
 
+    fn with_let_allowed(&mut self, allowed: bool, f: impl FnOnce(&mut Self, bool)) {
+        let old = mem::replace(&mut self.is_let_allowed, allowed);
+        f(self, old);
+        self.is_let_allowed = old;
+    }
+
+    /// Emits an error banning the `let` expression provided in the given location.
+    fn ban_let_expr(&self, expr: &'a Expr) {
```

Comment:
It's probably a better idea to move this to feature gating under e.g. `#![feature(let_expr)]` so that we can `::Let` more than what `let_chains` alone would allow. We could also move that, or the ast_validation logic, out of this PR and land it separately.